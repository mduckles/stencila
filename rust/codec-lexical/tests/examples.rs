use std::{fs::read_to_string, path::PathBuf};

use codec::{
    common::{eyre::Result, glob::glob},
    format::Format,
    EncodeOptions,
};

use codec_lexical::{decode, encode};
use common_dev::insta::{assert_json_snapshot, assert_snapshot, assert_yaml_snapshot};

/// Decode each example of a Lexical document and create JSON and Lexical snapshots
/// including snapshots for losses
#[test]
fn examples() -> Result<()> {
    let pattern = PathBuf::from(env!("CARGO_MANIFEST_DIR"))
        .join("tests/examples")
        .canonicalize()?
        .to_string_lossy()
        .to_string()
        + "/**/*";

    for path in glob(&format!("{pattern}.lexical"))?
        .chain(glob(&format!("{pattern}.koenig"))?)
        .flatten()
    {
        let format = Format::from_path(&path);
        let ext = format.extension();

        let name = path
            .file_name()
            .map(|name| name.to_string_lossy())
            .and_then(|name| name.strip_suffix(&[".", &ext].concat()).map(String::from))
            .expect("should have format suffix");

        let original = read_to_string(path)?;

        let (article, info) = decode(&original, None)?;

        assert_json_snapshot!(format!("{name}.json"), article);
        assert_yaml_snapshot!(format!("{name}.decode.losses"), info.losses);

        let (lexical, info) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format.clone()),
                compact: Some(false),
                ..Default::default()
            }),
        )?;

        assert_snapshot!(format!("{name}.{ext}"), lexical);
        assert_yaml_snapshot!(format!("{name}.encode.losses"), info.losses);

        let (lexical, ..) = encode(
            &article,
            Some(EncodeOptions {
                format: Some(format),
                standalone: Some(true),
                compact: Some(false),
                ..Default::default()
            }),
        )?;
        assert_snapshot!(format!("{name}.standalone.{ext}"), lexical);
    }

    Ok(())
}
