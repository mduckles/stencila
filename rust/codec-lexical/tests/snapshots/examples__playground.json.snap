---
source: rust/codec-lexical/tests/examples.rs
expression: article
snapshot_kind: text
---
{
  "type": "Article",
  "content": [
    {
      "type": "Heading",
      "level": 1,
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Welcome to the playground"
          }
        }
      ]
    },
    {
      "type": "QuoteBlock",
      "content": [
        {
          "type": "Paragraph",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "In case you were wondering what the black box at the bottom is â€“ it's the debug view, showing the current state of the editor. You can disable it by pressing on the settings control in the bottom-left of your screen and toggling the debug view setting."
              }
            }
          ]
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "The playground is a demo environment built with "
          }
        },
        {
          "type": "CodeInline",
          "code": {
            "string": "@lexical/react"
          }
        },
        {
          "type": "Text",
          "value": {
            "string": ". Try typing in "
          }
        },
        {
          "type": "Strong",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "some text"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " with "
          }
        },
        {
          "type": "Emphasis",
          "content": [
            {
              "type": "Text",
              "value": {
                "string": "different"
              }
            }
          ]
        },
        {
          "type": "Text",
          "value": {
            "string": " formats."
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Make sure to check out the various plugins in the toolbar. You can also use "
          }
        },
        {
          "type": "Text",
          "value": {
            "string": "LOST HashTag"
          }
        },
        {
          "type": "Text",
          "value": {
            "string": " or @-mentions too!"
          }
        }
      ]
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "If you'd like to find out more about Lexical, you can:"
          }
        }
      ]
    },
    {
      "type": "List",
      "items": [
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "Visit the "
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": "LOST Link"
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": " for documentation and more information."
                  }
                }
              ]
            }
          ]
        },
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "Check out the code on our "
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": "LOST Link"
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": "."
                  }
                }
              ]
            }
          ]
        },
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "Playground code can be found "
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": "LOST Link"
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": "."
                  }
                }
              ]
            }
          ]
        },
        {
          "type": "ListItem",
          "content": [
            {
              "type": "Paragraph",
              "content": [
                {
                  "type": "Text",
                  "value": {
                    "string": "Join our "
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": "LOST Link"
                  }
                },
                {
                  "type": "Text",
                  "value": {
                    "string": " and chat with the team."
                  }
                }
              ]
            }
          ]
        }
      ],
      "order": "Unordered"
    },
    {
      "type": "Paragraph",
      "content": [
        {
          "type": "Text",
          "value": {
            "string": "Lastly, we're constantly adding cool new features to this playground. So make sure you check back here when you next get a chance ðŸ™‚."
          }
        }
      ]
    }
  ]
}
